README.md
# iCloudLocationFetcher
Python script to fetch locations from iCloud and send the distance from home location to an url

More information will follow shortly


changelog.txt
Change history


Version 1.0.0, 2018-09-29
-------------------------
- ILF-43: Use Relative accuracy + Refresh interval
- ILF-42: Pyicloud can hang during authenticating due to infinite timeout value
- ILF-41: iOS11 sometimes gives inaccurate locations and fluctuates #43
- ILF-40: Enable override of sleep intervals and timeout settings
- ILF-38: Add optional config parameter to set home_radius
- ILF-36: In case of communication errors the retry should happen faster
- ILF-35: Always recreate connection after long sleep interval
- ILF-33: Add timeout to requests to iCloud server
- ILF-30: Timeout will exceed "Low Updates" endtime
- ILF-29: Warning is not logged with new set timeout
- ILF-27: Improve handling of PyiCloudAPIResponseError and reduce sleep interval

Version 0.7.0, 2017-11-12
-------------------------
- ILF-25: Improve communication with server to update
- ILF-19: Use less frequent update interval when home at night 
- ILF-21: iCloudLocationFetcher hangs resetting connections after a while
- ILF-20: Use cookie_directory to store the cookie in non-volotile memory
- ILF-15: Improve home detection when old locations are received
- ILF-14: Extend the directories to search for the conf file to include the iCloudLocationFetcher directory
- ILF-13: Running via init.d script requires shebang

Version 0.6.0, 2017-11-06
-------------------------
- ILF-10: Update interval reverts to long default interval when close to home
- ILF-8: Use adjustable sleep time based on devices next update request time
- ILF-7: Improve retry mechanism for outdated location fetching
- ILF-4: Use distance based update interval
- ILF-3: Store sensitive apple data in separate file

Version 0.5.0, 2017-11-03
-------------------------
- Change log line to correctly identify program as Locations2Domoticz
- For testing purposes make communicating with Domoticz optional
- Switch to requests package for Domoticz communication
- More robust error handling when communication with Domoticz
- Renamed project to iCloudLocationFetcher
- ILF-1: Make script Domoticz unaware but allow any URL to be called


twostep.py
import ConfigParser
import os
import sys
from pyicloud import PyiCloudService

# read configuration
config = ConfigParser.SafeConfigParser()
for loc in os.curdir, os.path.expanduser("~"):
    try:
        with open(os.path.join(loc, "iCloudLocationFetcher.conf")) as source:
            config.readfp(source)
    except IOError:
        pass

# read credentials
apple_creds_file = config.get('GENERAL', 'apple_creds_file')
try:
    with open(apple_creds_file) as f:
        appleid = f.readline().strip()
        password = f.readline().strip()
except IOError, e:
    print("Unable to read the apple credentials file '%s': %s" % (apple_creds_file, str(e)))
    sys.exit(1)

api = PyiCloudService(appleid, password, "~/.iCloudLocationFetcher")
if api.requires_2sa:
    import click
    print("Two-step authentication required. Your trusted devices are:")
    trusted_devices = api.trusted_devices
    for i, device in enumerate(trusted_devices):
        print("%s: %s" % (i, device.get('deviceName', "SMS to %s" % device.get('phoneNumber'))))

    device = click.prompt('Which device would you like to use?', default=0)
    device = trusted_devices[device]
    if not api.send_verification_code(device):
        print("Failed to send verification code")
        sys.exit(1)
    code = click.prompt('Please enter validation code')
    if not api.validate_verification_code(dict(), code):
        print("Failed to verify verification code")
        sys.exit(1)

# provide some information on the devices
devices = api.devices
for i, device in enumerate(devices):
    print("%s: type: %s, name: %s, location enabled: %s" % (i, device.content['deviceDisplayName'], device.content['name'], str(device.content['locationEnabled'])))


Location.py
# -*- coding: utf-8 -*-
import abc
import math
import time

ACCURATE_LIMIT_IN_M = 100  # if location accuracy is smaller than this then it is accurate
ACCURATE_LIMIT_WHEN_HOME_IN_M = 300  # if home, if accuracy is within this percentage of distance, then accurate enough
ACCURACY_TO_DISTANCE_PERCENTAGE = 20  # if accuracy is within this percentage of the distance, then accurate enough
MIN_ACCURACY_BETWEEN_LOCATIONS = 50  # if two locations are within this limit then they can be the same location

def distance_meters(origin, destination):
    lat1, lon1 = origin
    lat2, lon2 = destination
    radius_earth_km = 6371

    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = math.sin(dlat / 2) * math.sin(dlat / 2) + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) \
        * math.sin(dlon / 2) * math.sin(dlon / 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    d = int(round(radius_earth_km * c * 1000))

    return d


class Location(object):
    __metaclass__ = abc.ABCMeta
    home_position = None  # type: 'list'

    def __init__(self, latitude, longitude, accuracy, timestamp):
        self.latitude = latitude
        self.longitude = longitude
        self.accuracy = accuracy
        self.timestamp = timestamp

        self.distance_to_home = distance_meters((self.latitude, self.longitude), Location.home_position)
        # assume home if home_location is within accuracy
        if self.distance_to_home < self.accuracy:
            self.rounded_distance_km = 0.0
        else:
            self.rounded_distance_km = math.floor(self.distance_to_home / 100) / 10.0

    @classmethod
    def set_home_position(cls, home_position):
        cls.home_position = home_position

    def distance_to(self, other_location):
        return distance_meters((self.latitude, self.longitude), (other_location.latitude, other_location.longitude))

    def is_recent_enough(self, max_elapsed_time_in_s):
        return time.time() - self.timestamp < max_elapsed_time_in_s

    def is_accurate_enough(self):
        if self.accuracy < ACCURATE_LIMIT_IN_M:
            return True
        if self.rounded_distance_km == 0.0 and self.accuracy < ACCURATE_LIMIT_WHEN_HOME_IN_M:
            return True
        # accuracy is also fine if it is within 20% of the distance
        if self.accuracy < (self.distance_to_home / 100) * ACCURACY_TO_DISTANCE_PERCENTAGE:
            return True
        return False

    def is_home(self):
        return self.rounded_distance_km == 0.0

    def can_be_same_location(self, other_location):
        if other_location is None:
            return False
        if self.is_home() and other_location.is_home():
            return True
        distance_to_other = self.distance_to(other_location)
        return distance_to_other < max(self.accuracy, other_location.accuracy, MIN_ACCURACY_BETWEEN_LOCATIONS)

    def is_more_accurate(self, other_location):
        if other_location is None:
            return True
        if self.accuracy < other_location.accuracy:
            return True
        if self.accuracy == other_location.accuracy:
            return self.distance_to_home < other_location.distance_to_home
        return False

    def __str__(self):
        seconds_ago = time.time() - self.timestamp
        return "%d ± %dm, -%ds, %.1fkm" \
               % (self.distance_to_home, self.accuracy, seconds_ago, self.rounded_distance_km)


MonitorDevice.py
import abc
import datetime
import math
import requests
import time
from constants import ACTION_NEEDED_ERROR_SLEEP_TIME
from Location import Location

MIN_RETRIEVE_INTERVAL_IN_S = 15
MAX_RETRIEVE_INTERVAL_IN_S = 3600
DEFAULT_RETRIEVE_INTERVAL_IN_S = 300  # used when home
MAX_RETRIEVE_INTERVAL_IN_S_WHEN_NOT_MOVING_NEAR_HOME = 120 # used when near home and not moving for long time

RETRY_EXPONENTIAL_BASE_IN_S = 3

FARAWAY_CLOSEBY_DISTANCE_LIMIT_IN_KM = 15
SPEED_KM_PER_HOUR_FOR_FARAWAY = 120
SPEED_KM_PER_HOUR_FOR_CLOSEBY = 90

URL_DISTANCE_PARAM = "__DISTANCE__"
OUTDATED_LIMIT_IN_S = 60  # if icloud location timestamp is older than this, then retry
OUTDATED_LIMIT_IN_S_IN_HOME_PERIOD = 600  # if icloud location timestamp is older than this in home period, then retry


class MonitorDevice(object):
    __metaclass__ = abc.ABCMeta
    logger = None
    send_to_server = True

    def __init__(self, name, update_url):
        self.name = name
        self.update_url = update_url
        self.update_url_timestamp = 0
        self.home_period = None

        self.apple_device = None
        self.location_retrieved = None
        self.location_stored = None
        self.next_retrieve_timestamp = time.time()
        self.retrieve_retry_count = 0
        self.same_location_count = 0

    @classmethod
    def set_logger(cls, value):
        cls.logger = value

    @classmethod
    def set_send_to_server(cls, value):
        cls.send_to_server = value

    def get_apple_device(self):
        return self.apple_device

    def set_apple_device(self, value):
        self.apple_device = value

    def set_home_period(self, value):
        self.home_period = value

    def get_next_retrieve_timestamp(self):
        return self.next_retrieve_timestamp

    def should_update(self):
        return self.next_retrieve_timestamp < time.time()

    def is_moving(self):
        return self.same_location_count == 0

    def is_apple_device_ok(self):
        if self.apple_device is not None:
            if self.apple_device.content['locationEnabled']:
                return True
            else:
                self.logger.warn("Location disabled for '%s'. Next retry in %d seconds" % (
                    self.name, ACTION_NEEDED_ERROR_SLEEP_TIME))
        else:
            self.logger.warn("Device '%s' not found. Next retry in %d seconds" % (
                self.name, ACTION_NEEDED_ERROR_SLEEP_TIME))
        return False

    def update_location_retrieved(self):
        apple_location = self.apple_device.location()
        if apple_location is not None:
            self.logger.debug(apple_location)
            self.logger.debug("location: type=%s, finished=%s, horizontalAccuracy=%f" %
                              (apple_location['positionType'], apple_location['locationFinished'],
                               apple_location['horizontalAccuracy']))
            location_timestamp = apple_location['timeStamp'] / 1000
            accuracy = math.floor(apple_location['horizontalAccuracy'])
            self.location_retrieved = Location(apple_location['latitude'], apple_location['longitude'], accuracy,
                                               location_timestamp)
            return True
        else:
            self.logger.warn("Unable to get the location for device %s. Next retry in %d seconds" %
                             (self.name, ACTION_NEEDED_ERROR_SLEEP_TIME))
            return False

    def is_within_home_period(self):
        if self.home_period is None:
            return False

        # use next interval based on low_update_when_home_timespan
        now_dt = datetime.datetime.now()
        minutes_since_midnight = \
            math.floor((now_dt - now_dt.replace(hour=0, minute=0, second=0, microsecond=0)).total_seconds() / 60)
        if self.home_period[0] < self.home_period[1]:
            return self.home_period[0] < minutes_since_midnight < self.home_period[1]
        else:
            return minutes_since_midnight > self.home_period[0] or minutes_since_midnight < self.home_period[1]

    def send_to_update_url(self, old_distance_km, new_distance_km):
        url = self.update_url.replace(URL_DISTANCE_PARAM, str(new_distance_km))
        if self.send_to_server:
            self.logger.debug("About to update '%s' with '%s'" % (self.name, url))
            try:
                response = requests.get(url)
                self.update_url_timestamp = time.time()
                self.logger.debug("%s -> %s" % (url, response))
                if response.ok:
                    self.logger.info("Successfully updated distance of '%s' from %.1f to %.1f km" %
                                     (self.name, old_distance_km, new_distance_km))
                else:
                    self.logger.warn("Unable to update distance of '%s' using '%s'. Response: %s" %
                                     (self.name, url, response))
            except requests.ConnectionError, e:
                self.logger.error('Request failed %s - %s' % (url, e))
        else:
            self.update_url_timestamp = time.time()
            self.logger.info("Skipping sending update for '%s' to '%s'" % (self.name, url))

    def is_retrieved_location_better_and_message(self):
        '''
        Compares the location_retrieved to the location_stored and returns True, if
        - location_stored is None
        - location_retrieved overlaps with location_stored, but is more accurate
        - location_retrieved does not overlap location_stored
        :return: a boolean, with the comparision message
        '''
        if self.location_stored is None:
            self.same_location_count = 0
            return True, 'Setting initial location'

        if self.location_retrieved.can_be_same_location(self.location_stored):
            self.same_location_count += 1
            if self.location_retrieved.is_home():
                if self.location_retrieved.is_more_accurate(self.location_stored):
                    return True, 'Not moving but using more accurate location'
                else:
                    return False, 'Not moving'
            else:
                if self.location_retrieved.is_more_accurate(self.location_stored):
                    return True, 'Not moving (%d) but using more accurate location' % self.same_location_count
                else:
                    return False, 'Not moving (%d)' % self.same_location_count
        else:
            self.same_location_count = 0
            message = 'On the move'
            if self.location_retrieved.is_home():
                message = 'Arrived home'
            if self.location_stored.is_home():
                message = 'Left home'
            return True, message

    def update_retrieve_retry_count(self):
        recent_limit = OUTDATED_LIMIT_IN_S
        if self.is_within_home_period():
            recent_limit = OUTDATED_LIMIT_IN_S_IN_HOME_PERIOD

        if not self.location_retrieved.is_recent_enough(recent_limit):
            self.retrieve_retry_count += 1
            return 'Location is not recent enough'
        if not self.location_retrieved.is_accurate_enough():
            self.retrieve_retry_count += 1
            return 'Location is not accurate enough'
        self.retrieve_retry_count = 0
        return 'Location is acceptable'

    def update_next_retrieve_timestamp(self):
        # all went well, location is recent and accurate
        if self.retrieve_retry_count == 0:
            # if at home
            if self.location_retrieved.is_home():
                seconds_to_wait = self.calculate_seconds_to_sleep_when_home()
            else:  # not at home, so use distance based interval in range [min, max]
                if self.location_retrieved.rounded_distance_km > FARAWAY_CLOSEBY_DISTANCE_LIMIT_IN_KM:
                    regular_wait_seconds = int(self.location_retrieved.rounded_distance_km
                                               * 3600 / SPEED_KM_PER_HOUR_FOR_FARAWAY)
                else:
                    regular_wait_seconds = int(self.location_retrieved.rounded_distance_km
                                               * 3600 / SPEED_KM_PER_HOUR_FOR_CLOSEBY)

                if not self.is_moving() and regular_wait_seconds < MAX_RETRIEVE_INTERVAL_IN_S_WHEN_NOT_MOVING_NEAR_HOME:
                    seconds_to_wait = min(regular_wait_seconds +
                                          int(math.pow(RETRY_EXPONENTIAL_BASE_IN_S, self.same_location_count)),
                                        MAX_RETRIEVE_INTERVAL_IN_S_WHEN_NOT_MOVING_NEAR_HOME)
                else:
                    # use distance based interval
                    seconds_to_wait = regular_wait_seconds
        else:
            (retry_div, retry_mod) = divmod(self.retrieve_retry_count, 3)
            if retry_mod == 0:
                seconds_to_wait = min(retry_div * 60, DEFAULT_RETRIEVE_INTERVAL_IN_S)
            else:
                seconds_to_wait = MIN_RETRIEVE_INTERVAL_IN_S

        # adjust seconds to wait when moving, then we want to have a quicker update; 2/3
        if self.is_moving():
            seconds_to_wait = int(2 * seconds_to_wait / 3)
        self.next_retrieve_timestamp = time.time() + max(MIN_RETRIEVE_INTERVAL_IN_S, seconds_to_wait)

    def log_update_message(self, status_message, location_message):
        now = time.time()
        next_update = self.next_retrieve_timestamp - now
        next_message = 'Next update'
        if self.retrieve_retry_count > 0:
            next_message = 'Retry %d' % self.retrieve_retry_count
        elif not self.location_retrieved.is_home() and not self.is_moving():
            next_message = 'Next increased (%d) update' % self.same_location_count
        old_location_string = 'Unknown'
        if self.location_stored is not None:
            old_location_string = self.location_stored.__str__()
        self.logger.info("Device %s: %s. Old: (%s), New: (%s). %s. %s in %d seconds"
                         % (self.name, status_message, old_location_string, self.location_retrieved, location_message,
                            next_message, next_update))

    def retrieve_location_and_update(self):
        if self.is_apple_device_ok():
            if self.update_location_retrieved():
                (location_is_better, status_message) = self.is_retrieved_location_better_and_message()
                if location_is_better:
                    old_distance_km = -1.0
                    if self.location_stored is not None:
                        old_distance_km = self.location_stored.rounded_distance_km
                    if old_distance_km != self.location_retrieved.rounded_distance_km:
                        self.send_to_update_url(old_distance_km, self.location_retrieved.rounded_distance_km)
                location_message = self.update_retrieve_retry_count()
                self.update_next_retrieve_timestamp()
                self.log_update_message(status_message, location_message)
                if location_is_better:
                    self.location_stored = self.location_retrieved
            else:
                self.retrieve_retry_count = 0
                self.next_retrieve_timestamp = time.time() + ACTION_NEEDED_ERROR_SLEEP_TIME
        else:
            self.retrieve_retry_count = 0
            self.next_retrieve_timestamp = time.time() + ACTION_NEEDED_ERROR_SLEEP_TIME

    def calculate_seconds_to_sleep_when_home(self):
        if self.home_period is not None:
            # use next interval based on low_update_when_home_timespan
            now_dt = datetime.datetime.now()
            minutes_since_midnight = \
                math.floor((now_dt - now_dt.replace(hour=0, minute=0, second=0, microsecond=0)).total_seconds() / 60)
            if self.home_period[0] < self.home_period[1] \
                    and self.home_period[0] < minutes_since_midnight < self.home_period[1]:
                return min((self.home_period[1] - minutes_since_midnight) * 60, MAX_RETRIEVE_INTERVAL_IN_S)
            else:
                if minutes_since_midnight > self.home_period[0]:
                    return min(((24 * 60) - minutes_since_midnight + self.home_period[1]) * 60, MAX_RETRIEVE_INTERVAL_IN_S)
                if minutes_since_midnight < self.home_period[1]:
                    return min((self.home_period[1] - minutes_since_midnight) * 60, MAX_RETRIEVE_INTERVAL_IN_S)

        return DEFAULT_RETRIEVE_INTERVAL_IN_S


pyicloud/services/findmyiphone.py
import json
import sys

import six

from pyicloud.exceptions import PyiCloudNoDevicesException


class FindMyiPhoneServiceManager(object):
    """ The 'Find my iPhone' iCloud service

    This connects to iCloud and return phone data including the near-realtime
    latitude and longitude.

    """

    def __init__(self, service_root, session, params):
        self.session = session
        self.params = params
        self._service_root = service_root
        self._fmip_endpoint = '%s/fmipservice/client/web' % self._service_root
        self._fmip_refresh_url = '%s/refreshClient' % self._fmip_endpoint
        self._fmip_sound_url = '%s/playSound' % self._fmip_endpoint
        self._fmip_message_url = '%s/sendMessage' % self._fmip_endpoint
        self._fmip_lost_url = '%s/lostDevice' % self._fmip_endpoint

        self._devices = {}
        self.refresh_client()

    def refresh_client(self):
        """ Refreshes the FindMyiPhoneService endpoint,

        This ensures that the location data is up-to-date.

        """
        req = self.session.post(
            self._fmip_refresh_url,
            params=self.params,
            timeout=30,
            data=json.dumps(
                {
                    'clientContext': {
                        'fmly': True,
                        'shouldLocate': True,
                        'selectedDevice': 'all',
                    }
                }
            )
        )
        self.response = req.json()

        for device_info in self.response['content']:
            device_id = device_info['id']
            if device_id not in self._devices:
                self._devices[device_id] = AppleDevice(
                    device_info,
                    self.session,
                    self.params,
                    manager=self,
                    sound_url=self._fmip_sound_url,
                    lost_url=self._fmip_lost_url,
                    message_url=self._fmip_message_url,
                )
            else:
                self._devices[device_id].update(device_info)

        if not self._devices:
            raise PyiCloudNoDevicesException()

    def __getitem__(self, key):
        if isinstance(key, int):
            if six.PY3:
                key = list(self.keys())[key]
            else:
                key = self.keys()[key]
        return self._devices[key]

    def __getattr__(self, attr):
        return getattr(self._devices, attr)

    def __unicode__(self):
        return six.text_type(self._devices)

    def __str__(self):
        as_unicode = self.__unicode__()
        if sys.version_info[0] >= 3:
            return as_unicode
        else:
            return as_unicode.encode('ascii', 'ignore')

    def __repr__(self):
        return six.text_type(self)


class AppleDevice(object):
    def __init__(
        self, content, session, params, manager,
        sound_url=None, lost_url=None, message_url=None
    ):
        self.content = content
        self.manager = manager
        self.session = session
        self.params = params

        self.sound_url = sound_url
        self.lost_url = lost_url
        self.message_url = message_url

    def update(self, data):
        self.content = data

    def location(self):
        self.manager.refresh_client()
        return self.content['location']

    def status(self, additional=[]):
        """ Returns status information for device.

        This returns only a subset of possible properties.
        """
        self.manager.refresh_client()
        fields = ['batteryLevel', 'deviceDisplayName', 'deviceStatus', 'name']
        fields += additional
        properties = {}
        for field in fields:
            properties[field] = self.content.get(field)
        return properties

    def play_sound(self, subject='Find My iPhone Alert'):
        """ Send a request to the device to play a sound.

        It's possible to pass a custom message by changing the `subject`.
        """
        data = json.dumps({
            'device': self.content['id'],
            'subject': subject,
            'clientContext': {
                'fmly': True
            }
        })
        self.session.post(
            self.sound_url,
            params=self.params,
            data=data
        )

    def display_message(
        self, subject='Find My iPhone Alert', message="This is a note",
        sounds=False
    ):
        """ Send a request to the device to play a sound.

        It's possible to pass a custom message by changing the `subject`.
        """
        data = json.dumps(
            {
                'device': self.content['id'],
                'subject': subject,
                'sound': sounds,
                'userText': True,
                'text': message
            }
        )
        self.session.post(
            self.message_url,
            params=self.params,
            data=data
        )

    def lost_device(
        self, number,
        text='This iPhone has been lost. Please call me.',
        newpasscode=""
    ):
        """ Send a request to the device to trigger 'lost mode'.

        The device will show the message in `text`, and if a number has
        been passed, then the person holding the device can call
        the number without entering the passcode.
        """
        data = json.dumps({
            'text': text,
            'userText': True,
            'ownerNbr': number,
            'lostModeEnabled': True,
            'trackingEnabled': True,
            'device': self.content['id'],
            'passcode': newpasscode
        })
        self.session.post(
            self.lost_url,
            params=self.params,
            data=data
        )

    @property
    def data(self):
        return self.content

    def __getitem__(self, key):
        return self.content[key]

    def __getattr__(self, attr):
        return getattr(self.content, attr)

    def __unicode__(self):
        display_name = self['deviceDisplayName']
        name = self['name']
        return '%s: %s' % (
            display_name,
            name,
        )

    def __str__(self):
        as_unicode = self.__unicode__()
        if sys.version_info[0] >= 3:
            return as_unicode
        else:
            return as_unicode.encode('ascii', 'ignore')

    def __repr__(self):
        return '<AppleDevice(%s)>' % str(self)


iCloudLocationFetcher.py
#!/usr/bin/python

import ConfigParser
import logging
import os
import pyicloud
import requests
import signal
import sys
import time
from pyicloud.exceptions import PyiCloudAPIResponseError
from constants import ACTION_NEEDED_ERROR_SLEEP_TIME
from Location import Location
from MonitorDevice import MonitorDevice

MIN_SLEEP_TIME = 1
MAX_SLEEP_TIME = 3600
RECOVERABLE_ERROR_SLEEP_TIME = 60
MAX_SESSION_TIME = 1800  # icloud will respond with HTTP 450 if session is not used within this time

# Constants (Do not change)
SCRIPT_VERSION = "1.0.0"
SCRIPT_DATE = "2018-09-29"

# Global variables
keep_running = True
logger = None


# Initialisation
def init_logging(config):
    log_file = config.get('GENERAL', 'log_file')
    log_level_from_config = config.get('GENERAL', 'log_level')
    log_to_console = config.getboolean('GENERAL', 'log_to_console')
    log_level = logging.INFO
    if log_level_from_config == 'DEBUG':
        log_level = logging.DEBUG
    elif log_level_from_config == 'INFO':
        log_level = logging.INFO
    elif log_level_from_config == 'WARNING':
        log_level = logging.WARNING

    # logging to file
    logging.basicConfig(level=log_level, format='%(asctime)s %(levelname)-8s %(message)s',
                        filename=log_file, filemode='w')
    # logging to console
    if log_to_console:
        console = logging.StreamHandler()
        formatter = logging.Formatter('%(asctime)s %(levelname)-8s %(message)s')
        console.setFormatter(formatter)
        logging.getLogger('').addHandler(console)

    local_logger = logging.getLogger('locations2domoticz')

    return local_logger


def get_apple_device(devices, name):
    device = None
    i = 0
    while device is None and i < len(devices.keys()):
        if devices[i].content['name'] == name:
            device = devices[i]
        i += 1
    return device


# Main program
def main():
    global keep_running, logger

    # read configuration
    config = ConfigParser.SafeConfigParser({'low_updates_when_home': None,
                                            'send_to_server': "true",
                                            'home_radius': "0.0"})
    config_exists = False
    for loc in os.curdir, os.path.expanduser("~"), os.path.join(os.path.expanduser("~"), "iCloudLocationFetcher"):
        try:
            with open(os.path.join(loc, "iCloudLocationFetcher.conf")) as source:
                config.readfp(source)
                config_exists = True
        except IOError:
            pass

    if not config_exists:
        print("Error: Unable to find the 'iCloudLocationFetcher.conf' file. \n"
              "Put it in the current directory, in ~ or in ~/iCloudLocationFetcher.\n")
        sys.exit(1)

    logger = init_logging(config)

    logger.info("---")
    logger.info("iCloudLocationFetcher, v%s, %s" % (SCRIPT_VERSION, SCRIPT_DATE))

    def signal_handler(signal1, frame):
        global keep_running, logger
        logger.info('Signal received: %s from %s' % (signal1, frame))
        logger.debug('Setting keep_running to False')
        keep_running = False

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # read other configuration
    apple_creds_file = config.get('GENERAL', 'apple_creds_file')
    try:
        with open(apple_creds_file) as f:
            apple_id = f.readline().strip()
            apple_password = f.readline().strip()
    except IOError, e:
        logger.error("Unable to read the apple credentials file '%s': %s" % (apple_creds_file, str(e)))
        sys.exit(1)

    home_location_str = config.get('GENERAL', 'home_location')
    home_location = [float(x.strip()) for x in home_location_str.split(',')]
    Location.set_home_position(home_location)

    low_updates_when_home_timespan = None
    low_updates_when_home_str = config.get('GENERAL', 'low_updates_when_home')
    if low_updates_when_home_str is not None:
        low_updates_when_home_start_minutes = None
        low_updates_when_home_end_minutes = None
        low_updates_when_home_list = [x.strip() for x in low_updates_when_home_str.split('-')]
        if len(low_updates_when_home_list) == 2:
            hours_min_start_list = [int(x.strip()) for x in low_updates_when_home_list[0].split(':')]
            if len(hours_min_start_list) == 2:
                low_updates_when_home_start_minutes = 60 * hours_min_start_list[0] + hours_min_start_list[1]
            hours_min_end_list = [int(x.strip()) for x in low_updates_when_home_list[1].split(':')]
            if len(hours_min_end_list) == 2:
                low_updates_when_home_end_minutes = 60 * hours_min_end_list[0] + hours_min_end_list[1]

        if low_updates_when_home_start_minutes is None or low_updates_when_home_end_minutes is None:
            logger.warn("Invalid format of 'low_updates_when_home' parameter in config. Found '%s', but format should be '23:30-07:00'" % low_updates_when_home_str)
        else:
            logger.info("Low updates starting from %s (%d minutes) to %s (%d minutes)" % (low_updates_when_home_list[0], low_updates_when_home_start_minutes, low_updates_when_home_list[1], low_updates_when_home_end_minutes))
            low_updates_when_home_timespan = [low_updates_when_home_start_minutes, low_updates_when_home_end_minutes]

    send_to_server = config.getboolean('GENERAL', 'send_to_server')

    devices_to_monitor_str = config.get('GENERAL', 'devices_to_monitor')
    devices_to_monitor = devices_to_monitor_str.strip().split('\n')
    monitor_devices = []
    for device_to_monitor in devices_to_monitor:
        name_and_url = device_to_monitor.split(',')
        monitor_device = MonitorDevice(name_and_url[0], name_and_url[1])
        monitor_device.set_home_period(low_updates_when_home_timespan)
        monitor_devices.append(monitor_device)

    MonitorDevice.set_logger(logger)
    MonitorDevice.set_send_to_server(send_to_server)

    sleep_time = MIN_SLEEP_TIME
    icloud = None
    while keep_running:
        try:
            if icloud is None:
                icloud = pyicloud.PyiCloudService(apple_id, apple_password, "~/.iCloudLocationFetcher")
                if icloud.requires_2sa:
                    logger.error("Two-step authentication required. Please run twostep.py")
                    sleep_time = ACTION_NEEDED_ERROR_SLEEP_TIME
                    icloud = None
                else:
                    sleep_time = MIN_SLEEP_TIME
                    icloud_devices = icloud.devices
                    # set corresponding devices
                    for monitor_device in monitor_devices:
                        logger.info("Searching for '%s' in iCloud devices" % monitor_device.name)
                        apple_device = get_apple_device(icloud_devices, monitor_device.name)
                        if apple_device is not None:
                            monitor_device.set_apple_device(apple_device)
                            logger.info("Found iCloud device '%s'" % str(apple_device))
                        else:
                            logger.warn("No iCloud device found with name '%s'" % monitor_device.name)

            if icloud is not None:
                now = time.time()
                next_sleep_time = MAX_SLEEP_TIME
                for monitor_device in monitor_devices:
                    if monitor_device.should_update():
                        monitor_device.retrieve_location_and_update()
                        next_update = int(monitor_device.get_next_retrieve_timestamp() - now)
                        next_sleep_time = min(next_sleep_time, next_update + MIN_SLEEP_TIME)
                    else:
                        next_update = int(monitor_device.get_next_retrieve_timestamp() - now)
                        logger.debug("Update not needed yet for '%s'. Next update in %d seconds" % (monitor_device.name, next_update))
                        next_sleep_time = min(next_sleep_time, next_update + MIN_SLEEP_TIME)
                sleep_time = next_sleep_time

        except PyiCloudAPIResponseError as e:
            logger.warn("PyiCloudAPIResponseError: {0}. Sleeping for {1} seconds".format(str(e), str(RECOVERABLE_ERROR_SLEEP_TIME)))
            icloud = None
            sleep_time = RECOVERABLE_ERROR_SLEEP_TIME
        except requests.exceptions.ConnectionError as e:
            logger.warn("ConnectionError: {0}. Sleeping for {1} seconds".format(str(e), str(RECOVERABLE_ERROR_SLEEP_TIME)))
            icloud = None
            sleep_time = RECOVERABLE_ERROR_SLEEP_TIME
        except requests.Timeout as e:
            logger.warn("Timout: {0}. Sleeping for {1} seconds".format(str(e), str(RECOVERABLE_ERROR_SLEEP_TIME)))
            icloud = None
            sleep_time = RECOVERABLE_ERROR_SLEEP_TIME
        except:
            logger.exception("Unexpected exception. Sleeping for {0} seconds".format(ACTION_NEEDED_ERROR_SLEEP_TIME))
            icloud = None
            sleep_time = ACTION_NEEDED_ERROR_SLEEP_TIME

        if sleep_time >= MAX_SESSION_TIME:
            icloud = None
            logger.debug("Sleeping for %d seconds and resetting icloud connection" % sleep_time)
        else:
            logger.debug("Sleeping for %d seconds" % sleep_time)
        time.sleep(sleep_time)


if __name__ == '__main__':
    main()
